src/Language/Fixpoint/Smt/Interface.hs

```diff
@@ -240,12 +236,17 @@ negativeP
 makeContext :: Config -> FilePath -> IO Context
 --------------------------------------------------------------------------
 makeContext cfg f
-  = do createDirectoryIfMissing True $ takeDirectory smtFile
-       hLog <- openFile smtFile WriteMode
-       hSetBuffering hLog $ BlockBuffering $ Just $ 1024 * 1024 * 64
-       me   <- makeContext' cfg $ Just hLog
+  = do mb_hLog <- if not (save cfg) then pure Nothing else do
+           createDirectoryIfMissing True $ takeDirectory smtFile
+           hLog <- openFile smtFile WriteMode
+           hSetBuffering hLog $ BlockBuffering $ Just $ 1024 * 1024 * 64
+           return $ Just hLog
+       me   <- makeContext' cfg mb_hLog
        pre  <- smtPreamble cfg (solver cfg) me
-       mapM_ (\l -> SMTLIB.Backends.command_ (ctxSolver me) l >> BS.hPutBuilder hLog l >> LBS.hPutStr hLog "\n") pre
+       mapM_ (SMTLIB.Backends.command_ (ctxSolver me)) pre
+       forM_ mb_hLog $ \hLog -> forM_ pre $ \l -> do
+           BS.hPutBuilder hLog l
+           LBS.hPutStr hLog "\n"
```

Comment:
If there is no particular reason for the current arrangement, maybe it is worth preserving the order in which commands are sent one at a time to both the SMT solver and the log, in case the SMT solver crashes.

Reply:
Yeah, I was debating with myself on whether I should be conservative and do that, or write somewhat simpler code. So now its 2:1 in favor of being conservative :), hence I updated the PR.